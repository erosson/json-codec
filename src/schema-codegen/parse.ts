import * as S from 'json-schema'

export type SchemaDefinition = { name: string, body: SchemaBody }
export type SchemaBody = SchemaSimpleType | SchemaObject | SchemaArray | SchemaRef
export type SchemaSimpleType = { type: 'string' } | { type: 'number' } | { type: 'boolean' } | { type: 'null' }
export type SchemaObject = { type: 'object', properties: SchemaProp[] }
export type SchemaProp = SchemaDefinition
export type SchemaArray = { type: 'array', items: SchemaBody }
export type SchemaRef = { type: 'ref', path: string }
export type Schema = SchemaDefinition[]

function refName(ref: SchemaRef): string | null {
    if (ref.path.startsWith('#/definitions/')) {
        return ref.path.replace('#/definitions/', '')
    }
    return null
}
export function parse(schema: S.JSONSchema7): Schema {
    return [parseDefinition(schema)]
        .concat(Object.entries(schema.definitions ?? {}).map(([name, def]) => parseDefinition(def, name)))
        .filter(d => !!d) as SchemaDefinition[]
}
function parseDefinition(def: S.JSONSchema7Definition, name?: string): SchemaDefinition | null {
    if (typeof def !== 'boolean') {
        name = name ?? def.title
        if (name) {
            const body = parseBody(def)
            if (body) {
                return { name, body }
            }
        }
    }
    return null
}

function parseBody(def: S.JSONSchema7): SchemaBody | null {
    if (def.$ref) {
        return { type: 'ref', path: def.$ref }
    }
    switch (def.type) {
        case 'string':
        case 'number':
        case 'boolean':
        case 'null':
            return { type: def.type }
        case 'integer':
            return { type: 'number' }
        case 'object':
            return parseObject(def)
        case 'array':
            return parseArray(def)
        case undefined:
        default:
            return null
    }
}

function parseArray(def: S.JSONSchema7): SchemaArray | null {
    if (def.items && typeof def.items === 'object' && !Array.isArray(def.items)) {
        const body = parseBody(def.items)
        if (body) {
            return { type: 'array', items: body }
        }
    }
    return null
}
function parseObject(def: S.JSONSchema7): SchemaObject {
    return {
        type: 'object',
        properties: Object.entries(def.properties ?? {}).map(([key, body]) => {
            return parseProp(body, key)
        })
            .filter(p => !!p) as SchemaProp[]
    }
}
function parseProp(def: S.JSONSchema7Definition, name: string): SchemaProp | null {
    return parseDefinition(def, name)
}