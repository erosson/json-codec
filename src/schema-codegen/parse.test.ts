import * as P from './parse'
import schema from '../FreeCBT.schema.json'

test('parse DistortionID', () => {
    expect(P.parse({ ...schema.definitions.DistortionID, title: 'DistortionID' } as any)).toMatchSnapshot()
})
test('parse Distortion', () => {
    const parsed = P.parse({ ...schema.definitions.Distortion, title: 'Distortion' } as any)
    expect(parsed[0].body.type === 'object')
    expect(new Set((parsed[0].body as P.SchemaObject).properties.map(p => p.name))).toEqual(new Set(Object.keys(schema.definitions.Distortion.properties)))
    expect(parsed).toMatchSnapshot()
})
test('parse ThoughtID', () => {
    expect(P.parse({ ...schema.definitions.ThoughtID, title: 'ThoughtID' } as any)).toMatchSnapshot()
})
test('parse Thought', () => {
    const parsed = P.parse({ ...schema.definitions.Thought, title: 'Thought' } as any)
    expect(parsed[0].body.type === 'object')
    expect(new Set((parsed[0].body as P.SchemaObject).properties.map(p => p.name))).toEqual(new Set(Object.keys(schema.definitions.Thought.properties)))
    expect(parsed).toMatchSnapshot()
})
