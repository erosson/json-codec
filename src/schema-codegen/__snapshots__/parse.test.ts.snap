// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse Distortion 1`] = `
[
  {
    "body": {
      "properties": [
        {
          "body": {
            "path": "#/definitions/DistortionID",
            "type": "ref",
          },
          "name": "slug",
        },
        {
          "body": {
            "items": {
              "type": "string",
            },
            "type": "array",
          },
          "name": "emoji",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "labelKey",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "descriptionKey",
        },
      ],
      "type": "object",
    },
    "name": "Distortion",
  },
]
`;

exports[`parse DistortionID 1`] = `
[
  {
    "body": {
      "type": "string",
    },
    "name": "DistortionID",
  },
]
`;

exports[`parse Thought 1`] = `
[
  {
    "body": {
      "properties": [
        {
          "body": {
            "type": "string",
          },
          "name": "automaticThought",
        },
        {
          "body": {
            "items": {
              "path": "#/definitions/DistortionID",
              "type": "ref",
            },
            "type": "array",
          },
          "name": "cognitiveDistoritions",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "challenge",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "alternativeThought",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "createdAt",
        },
        {
          "body": {
            "type": "string",
          },
          "name": "updatedAt",
        },
        {
          "body": {
            "path": "#/definitions/ThoughtID",
            "type": "ref",
          },
          "name": "uuid",
        },
      ],
      "type": "object",
    },
    "name": "Thought",
  },
]
`;

exports[`parse ThoughtID 1`] = `
[
  {
    "body": {
      "type": "string",
    },
    "name": "ThoughtID",
  },
]
`;
